File: loader/src/main/java/github/dqw4w9wgxcq/botapi/loader/ScriptManager.java
package github.dqw4w9wgxcq.botapi.loader;

import lombok.Getter;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

import java.io.File;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.*;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;
import java.util.stream.Collectors;

@Slf4j
public class ScriptManager {
    private final File SCRIPTS_DIR = new File(new File(RealUserHome.getUserHome(), "runelite-bot"), "scripts");

    @Getter
    private final ScriptThread scriptThread = new ScriptThread();

    private final ClassLoader classLoader;

    public ScriptManager(ClassLoader classLoader) {
        this.classLoader = classLoader;
        log.info("classloader:{}", classLoader);

        scriptThread.start();
    }

    public void startScript(Class<? extends IBotScript> scriptClass) {
        IBotScript activeScript = scriptThread.getScript();
        if (activeScript != null) {
            log.info("script is running already");
            return;
        }

        IBotScript script;
        try {
            script = scriptClass.getDeclaredConstructor().newInstance();
        } catch (Exception e) {
            log.warn("exception initing script", e);
            return;
        }

        boolean accepted = scriptThread.offer(script);

        if (!accepted) {
            log.warn("script not accepted by scriptThread");
            return;
        }

        log.info("started script" + script.getClass().getSimpleName());
    }

    public void startScript(String scriptName) {
        for (Class<? extends IBotScript> scriptClass : loadScripts()) {
            if (scriptClass.getAnnotation(ScriptMeta.class).value().trim().equalsIgnoreCase(scriptName.trim())) {
                startScript(scriptClass);
                return;
            }
        }

        throw new IllegalArgumentException("script not found: " + scriptName);
    }

    public void stopScript() {
        IBotScript activeScript = scriptThread.getScript();
        if (activeScript == null) {
            log.info("no script running");
        } else {
            log.info("asking script {} to stop", activeScript.getClass().getSimpleName());
            activeScript.stopLooping();
        }
    }

    public List<Class<? extends IBotScript>> loadScripts() {
        List<Class<? extends IBotScript>> out = new ArrayList<>(loadScriptsFromDir(SCRIPTS_DIR));

        File desktopScriptJarFile = new File(new File(RealUserHome.getUserHome(), "Desktop"), "script.jar");
        if (desktopScriptJarFile.exists()) {
            out.addAll(loadScriptsFromFile(desktopScriptJarFile));
        } else {
            log.info("desktop script jar not found");
        }

        return out;
    }

    private List<Class<? extends IBotScript>> loadScriptsFromDir(File dir) {
        File[] files = dir.listFiles(pathname -> !pathname.isDirectory() && pathname.getName().endsWith(".jar"));

        if (files == null) {
            log.warn("no script directory");
            return Collections.emptyList();
        }

        List<Class<? extends IBotScript>> out = new ArrayList<>();
        for (File file : files) {
            out.addAll(loadScriptsFromFile(file));
        }

        return out;
    }

    @SneakyThrows
    @SuppressWarnings("unchecked")
    private List<Class<? extends IBotScript>> loadScriptsFromFile(File file) {
        List<Class<? extends IBotScript>> out = new ArrayList<>();
        long startTime = System.currentTimeMillis();

        try (JarFile jar = new JarFile(file)) {
            try (URLClassLoader ucl = new URLClassLoader(new URL[]{file.toURI().toURL()}, getClass().getClassLoader())) {
                List<String> scriptFileNames = Arrays.stream(ucl.getURLs()).map(URL::getFile).collect(Collectors.toList());
                log.info("script jars:" + String.join(", ", scriptFileNames));

                Enumeration<JarEntry> entries = jar.entries();
                while (entries.hasMoreElements()) {
                    JarEntry entry = entries.nextElement();
                    String name = entry.getName();

                    if (name.contains("module-info")) {//java 9 (kotlin dep has it)
                        continue;
                    }

                    if (name.endsWith(".class")) {
                        name = name.substring(0, name.length() - ".class".length());
                        name = name.replace('/', '.');
                        log.debug("loading class with name: {}", name);

                        Class<?> clazz = ucl.loadClass(name);

                        if (clazz.getAnnotation(ScriptMeta.class) != null && IBotScript.class.isAssignableFrom(clazz)) {
                            out.add((Class<? extends IBotScript>) clazz);
                        }
                    }
                }
            }
        }

        long time = System.currentTimeMillis() - startTime;
        log.info(
                "loaded scripts {} in {}ms",
                out.stream().map(c -> c.getAnnotation(ScriptMeta.class).value()).collect(Collectors.toList()),
                time
        );

        return out;
    }
}

File: loader-api/src/main/java/github/dqw4w9wgxcq/botapi/loader/ScriptManager.kt
package github.dqw4w9wgxcq.botapi.loader

import org.slf4j.LoggerFactory
import java.io.File
import java.lang.reflect.Modifier
import java.net.URLClassLoader
import java.util.jar.JarFile

class ScriptManager {
    private val log = LoggerFactory.getLogger(ScriptManager::class.java)
    val scriptsDir = File(File(System.getProperty("user.home"), "runelite-bot"), "scripts")

    private val scriptThread = ScriptThread().also { it.start() }

    /**
     * @return true if new script was successfully started, false if script was already running or failed to start
     */
    fun startScript(scriptClass: Class<out IBotScript>): Boolean {
        val activeScript = scriptThread.activeScript
        if (activeScript != null) {
            log.info("script is running already")
            return false
        }

        val script: IBotScript = try {
            scriptClass.getDeclaredConstructor().newInstance()
        } catch (e: Exception) {
            log.warn("script init", e)
            return false
        }

        val accepted = scriptThread.offer(script)

        if (!accepted) {
            log.warn("script $script not accepted by scriptThread")
            return false
        }

        log.info("started script" + script::class.java.simpleName)
        return true
    }

    /**
     * @return true if no script running
     */
    fun stopScript(): Boolean {
        val activeScript = scriptThread.activeScript
        return if (activeScript == null) {
            log.info("no script running")
            true
        } else {
            log.info("asking script to stop")
            activeScript.stopLooping()
            false
        }
    }

    fun loadScripts(): List<Class<out IBotScript>> {
        return buildList {
            addAll(loadScriptsFromDir(scriptsDir))

            val desktopScriptJarFile = File(System.getProperty("user.home")).resolve("Desktop").resolve("script.jar")
            if (desktopScriptJarFile.exists()) {
                addAll(loadScriptsFromFile(desktopScriptJarFile))
            } else {
                log.info("desktop script jar not found")
            }
        }
    }

    private fun loadScriptsFromDir(dir: File): List<Class<out IBotScript>> {
        return buildList {
            val files = dir.listFiles { pathname: File -> !pathname.isDirectory && pathname.name.endsWith(".jar") }
            if (files == null) {
                log.warn("no script directory")
                return emptyList()
            }

            for (file in files) {
                addAll(loadScriptsFromFile(file))
            }
        }
    }

    private fun loadScriptsFromFile(file: File): List<Class<out IBotScript>> {
        val out: MutableList<Class<out IBotScript>> = ArrayList()
        val startTime = System.currentTimeMillis()

        JarFile(file).use { jar ->
            URLClassLoader(arrayOf(file.toURI().toURL())).use { ucl ->
                val elems = jar.entries()
                while (elems.hasMoreElements()) {
                    val entry = elems.nextElement()
                    var name = entry.name
                    if (name.contains("module-info")) continue//idk lol
                    if (name.endsWith(".class")) {
                        name = name.substring(0, name.length - ".class".length)
                        name = name.replace('/', '.')
                        val clazz = ucl.loadClass(name)
                        if (clazz.getAnnotation(ScriptMeta::class.java) != null && IBotScript::class.java.isAssignableFrom(
                                clazz
                            ) && !Modifier.isAbstract(clazz.modifiers)
                        ) {
                            @Suppress("UNCHECKED_CAST") out.add(clazz as Class<out IBotScript>)
                        }
                    }
                }
            }
        }

        log.info("loaded scripts[${out.joinToString { it.getAnnotation(ScriptMeta::class.java).value }}] from file: $file in ${System.currentTimeMillis() - startTime}ms")

        return out
    }
}
